lib/common/format.cc

```diff
@@ -0,0 +1,616 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "format.h"
+
+namespace Fortran::common {
+
+template<typename CHAR> CHAR FormatValidator<CHAR>::NextChar() {
+  for (++cursor_; cursor_ < end_; ++cursor_) {
+    if (*cursor_ != ' ') {
+      return toupper(*cursor_);
+    }
+  }
+  cursor_ = end_;
+  return ' ';
+};
+
+template<typename CHAR> CHAR FormatValidator<CHAR>::LookAheadChar() {
+  for (laCursor_ = cursor_ + 1; laCursor_ < end_; ++laCursor_) {
+    if (*laCursor_ != ' ') {
+      return toupper(*laCursor_);
+    }
+  }
+  laCursor_ = end_;
+  return ' ';
+};
+
+template<typename CHAR> void FormatValidator<CHAR>::Advance(TokenKind tk) {
+  cursor_ = laCursor_;
+  token_.set_kind(tk);
+};
+
+template<typename CHAR> void FormatValidator<CHAR>::NextToken() {
+  // At entry, cursor_ points before the start of the next token.
+  // At exit, cursor_ points to last CHAR of token_.
+
+  CHAR c{NextChar()};
+  token_.set_kind(TokenKind::None);
+  token_.set_offset(cursor_ - format_);
+  token_.set_length(1);
+  if (c == '_' && integerValue_ >= 0) {  // C1305, C1309, C1310, C1312, C1313
+    AppendError("Kind parameter '_' character in format expression");
+  }
+  integerValue_ = -1;
+
+  switch (c) {
+  case '0':
+  case '1':
+  case '2':
+  case '3':
+  case '4':
+  case '5':
+  case '6':
+  case '7':
+  case '8':
+  case '9': {
+    int64_t lastValue;
+    const CHAR *lastCursor;
+    integerValue_ = 0;
+    bool overflow{false};
+    do {
+      lastValue = integerValue_;
+      lastCursor = cursor_;
+      integerValue_ = 10 * integerValue_ + c - '0';
```

Comment:
does `c - '0'` work the same for all `CHAR` types?

Reply:
Yes, I  believe so.  I don't know how to exercise this code for kind=2/4 characters with an external program, but I have invoked it a few times with forced strings.  Those tests look good, including tests with integer values.